File: leetcode/78.search-a-2d-matrix-ii/suhy-reduction.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
const searchMatrix = (matrix, target) => {
    let row = matrix.length - 1
    let col = 0
    while (row >= 0 && col < matrix[0].length) {
        if (matrix[row][col] === target) {
            return true;
        } else if (matrix[row][col] < target) {
            col += 1;
        } else {
            row -= 1;
        }
    }
    return false;
    
};

File: leetcode/78.search-a-2d-matrix-ii/suhy-reduction.rb
# @param {Integer[][]} matrix
# @param {Integer} target
# @return {Boolean}
class SpaceReduction
    def initialize(matrix, target)
        @matrix = matrix
        @target = target
    end
    
    def search_matrix
        row = @matrix.length - 1
        col = 0
        while row >= 0 && col < @matrix[0].length
            if @matrix[row][col] == @target
                return true
            elsif @matrix[row][col] < @target
                col += 1
            elsif @matrix[row][col] > @target
                row -= 1
            end
        end
        return false
    end
end

def search_matrix(matrix, target)
    SpaceReduction.new(matrix, target).search_matrix
end
